File: leetcode/331_VerifyPreorderSerializationOfBinaryTree/lc331.py
class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        l = preorder.split(',')
        children = []
        for i in range(len(l)):
            if l[i] == '#':
                children.append(0)
            else:
                children.append(2)
            for j in range(i):
                if children[i-1-j] > 0:
                    children[i-1-j] -= 1
                    break
                if j == i-1:
                    return False
        if sum(children) > 0:
            return False
        else:
            return True

File: leetcode/331_VerifyPreorderSerializationOfBinaryTree/lc331.cpp
class Solution {
public:
    bool isValidSerialization(string preorder) {
        int degree = 1;
        int i = 0;
        int n = preorder.length();
        while (i < n) {
            if (preorder[i] != ',') {
                degree--;
                if (degree < 0) return false; // in_degree == out_degree
                if (preorder[i] != '#') {
                    while (isdigit(preorder[i])) i++;
                    degree += 2;
                } else {
                    i++;
                }
            }
             else {
                i++;
            }
        }
        return degree == 0;
    }
};
